### PR TITLE
Add Python case versioning capability

### LINKED ISSUE #106: Add capability to version cases
Add capability to add a `case_version` tag depending on parameter values as in https://github.com/ursacomputing/arrowbench/pull/105 . This only includes versioning Python benchmarks, not R ones, whose versions will be read and passed through from their JSON in a different story.

### PR DESCRIPTION
**Pull request by alistaire47**
Closes #106; parallel to https://github.com/ursacomputing/arrowbench/pull/105. Adds the ability to version cases by overwriting a `version_cases()` method that takes in a case tuple and returns `None` (no versioning; continue current behavior) or an integer, which will be appended to `tags` as `case_version`.

Also adds a pile of type hints to `_benchmark.py` because I wrote them out before when trying to understand data flow.

Also changes the TPC-H benchmark to allow for smaller scale factors and uses them in tests so they don't take half an hour on CI.

### CODE REVIEW

**Comment by alistaire47** [comment]
Hmm, TPC-H still took 23 min: 

```
1278.24s call     benchmarks/tests/test_tpch_benchmark.py::test_benchmark_r
```
Not sure why; it's very fast locally

**Comment by jonkeane** [comment]
> Hmm, TPC-H still took 23 min

This is probably due to needing to compile duckdb from source to get the tpc-h extension. We should check and see if the extension loading mechanism is finally working or not — I know that there have been a few PRs that were needed to get the extension loading working, but I'm not certain those have merged + released + sent to CRAN.

File: benchmarks/_example_benchmarks.py [hunk 1]
```diff
@@ -79,6 +79,9 @@ class CasesBenchmark(_benchmark.Benchmark):
         ("10", "2"),
     )
 
+    def version_case(self, case: tuple):
+        return 2 * int(case[0]) + int(case[1])
```
Comment (jonkeane) [inline benchmarks/_example_benchmarks.py:83]: Am I following correctly that this will end up being 30, 14, and 22?

Comment (jonkeane) [reply benchmarks/_example_benchmarks.py:83]: _nods_ we might want to comment that reasoning here to remind our future selves what's going on. My first read of it was that we were making claims about how the version numbers should be derivable from the contents of the case like this, when really they are labels we apply to particular cases.

Comment (alistaire47) [reply benchmarks/_example_benchmarks.py:83]: Hm, yeah, maybe for reference purposes it's more useful to put a lookup dict here, e.g. something like 

``` python
case_version_dict = {
    ("10", "10"): 1,
    ("2", "10"): 2,
    ("10", "2"): 1,
}
case_version = case_version_dict[case]
```

File: benchmarks/_example_benchmarks.py [hunk 2]
```diff
@@ -79,6 +79,17 @@ class CasesBenchmark(_benchmark.Benchmark):
         ("10", "2"),
     )
 
+    def version_case(self, case: tuple):
+        # A lookup like this might be a common way to store versions, but any logic here
+        # (e.g. `2 if case[0] == "2" else 1`) is fine!
```
Comment (jonkeane) [inline benchmarks/_example_benchmarks.py:84]: 😍

**Review by jonkeane — APPROVED** [review]
Love the expanded comments in the example

**Pull request merged**